Fix settings screen layout on iOS and add soft back button

zbettenbuk opened the pull request.

saghul commented on react/features/app-settings/components/AppSettings.native.js:
```diff
@@ -39,7 +41,11 @@ class AppSettings extends AbstractAppSettings {
                 presentationStyle = 'fullScreen'
                 style = { styles.modal }
                 visible = { this.props._visible }>
+                <StatusBar hidden = { true } />
```
Why do we want this though? I think we shouldn't hide it.

saghul commented:
There needs to be a top margin equal to the one on the welcome page, otherwise it's broken for the iPhone X:

<img width="545" alt="screen shot 2018-01-08 at 13 27 14" src="https://user-images.githubusercontent.com/317464/34670696-b7697316-f477-11e7-8b9e-6b16b96d2ddc.png">

saghul commented:
Also, the settings page doesn't seem to handle rotation, whereas the welcome page does.

zbettenbuk commented:
@saghul The settings screen now uses a custom implementation of paddings around header and content minimising the blank area on every phone, also focusing on iPhone X. It doesn't support the Essential Phone as I didn't have a method to detect it and nor device to test. But I think we can live with that for now, especially as the top notch of that will likely not interfere with the "Settings" label, being quite narrow.

In long term we need a more generic solution as we discussed, but I'm not a big fan of anything that renders too much static-empty areas where not really needed.

zbettenbuk commented:
@saghul Also: I fixed the crash around the URL validation, and removed the overuse of component state in the abstract logic (hence the removed type and some functions).

zbettenbuk commented:
![image](https://user-images.githubusercontent.com/9974975/34741884-7ee4b952-f584-11e7-9a28-c10126039a2b.png)
![image](https://user-images.githubusercontent.com/9974975/34741891-869a9b6c-f584-11e7-8a25-4fd382cf4c25.png)

saghul commented on react/features/app-settings/components/AbstractAppSettings.js:
```diff
@@ -304,8 +173,19 @@ export class AbstractAppSettings extends Component<Props, State> {
+
+    console.log('ZB ::: New profile', JSON.stringify(_profile));
+
+    if (typeof _profile.defaultURL !== 'string') {
```
why do we need this? why not leave it as undefined? right now it shows in "black" as if the user typed it, IMHO it should remain undefined.

saghul commented on react/features/app-settings/components/AppSettings.native.js:
```diff
@@ -23,6 +28,17 @@ import { translate } from '../../base/i18n';
  */
 class AppSettings extends AbstractAppSettings {
 
+    /**
```
style: unneeded extra newline

zbettenbuk replied:
I insist 😆  I my world class definitions need an extra line after

saghul replied:
Please look around and be consistent.

saghul commented on react/features/app-settings/components/AppSettings.native.js:
```diff
@@ -36,42 +52,49 @@ class AppSettings extends AbstractAppSettings {
                 onRequestClose = { this._onRequestClose }
-                presentationStyle = 'fullScreen'
-                style = { styles.modal }
+                presentationStyle = 'overFullScreen'
```
Why is this needed? According to the docs it's the same as fullScreen, plus transparency, which we are not using: https://facebook.github.io/react-native/docs/modal.html#presentationstyle It also seems to be the default, so can we just remove the prop altogether?

zbettenbuk replied:
It's needed for the underlying view to be re-layout'ed when rotated, so then this screen receives the aspect ration change. It doesn't work with "fullScreen"

saghul replied:
I commented it out and it worked for me. Looks like `supportedOrientations` is the only required thing. Can you please double-check?

zbettenbuk replied:
Re-checked, the orientation does work, but the screen doesn't get re-layout'd and the paddings will be outdated. Please test it on an iPhone (the most visible it is on an iPhone X). Open the settings screen in portrait, then rotate the device. The header will still have the safety offset (where it shouldn't) and the side won't have it (while it should). Or is it only me?
![image](https://user-images.githubusercontent.com/9974975/34768969-a23d9e32-f5fc-11e7-8139-38b12bbe914b.png)

saghul replied:
Isn't the aspectRatio a prop? If so, it would trigger a re-render.

zbettenbuk replied:
Yes, bit it doesn't change

saghul replied:
I tested this and (I guess because this is a Modal) our aspect ratio detector doesn't kick in, and thus there is no new props when we rotate. Sigh.

saghul replied:
I "solved" it as follows:

```
    componentWillMount() {
        this._orientationListener = DeviceEventEmitter.addListener(
            'namedOrientationDidChange', () => {
                console.log('ORIENTATION CHANGED');
                this.forceUpdate();
            },
        );
    }

    componentWillUnmount() {
        this._orientationListener.remove();
    }
```

I'm thinking we might want to use that for the aspect ratio detector...

zbettenbuk replied:
I leave it to your PR then to fix the detector as we agreed.

saghul commented on react/features/app-settings/components/AppSettings.native.js:
```diff
@@ -36,42 +52,49 @@ class AppSettings extends AbstractAppSettings {
+                presentationStyle = 'overFullScreen'
+                supportedOrientations = { [
+                    'portrait',
+                    'landscape'
```
style: alpha-sort

saghul commented on react/features/app-settings/components/AppSettings.native.js:
```diff
@@ -36,42 +52,49 @@ class AppSettings extends AbstractAppSettings {
-                            placeholder = 'https://jitsi.example.com'
-                            value = { this.state.serverURL } />
+                            placeholder = { this.props._configDefaultURL }
+                            value = { this.props._profile.defaultURL } />
```
we seem to have a mix of serverURL and defaultURL. Can we make that consistent?

saghul commented on react/features/app-settings/components/AppSettings.native.js:
```diff
@@ -80,20 +103,43 @@ class AppSettings extends AbstractAppSettings {
                             }
-                            value = { this.state.startWithVideoMuted } />
+                            value = {
+                                this.props._profile.startWithVideoMuted
```
general observation: I think it's worth it to deconstruct the props / profile to avoid long lines.

saghul commented on react/features/app-settings/components/BackButton.native.js:
```diff
@@ -0,0 +1,59 @@
+/**
+ * A component rendering a back button that looks native on both platforms.
+ */
+export default class BackButton extends Component<Props, *> {
```
ditto for the *

saghul commented on react/features/app-settings/components/BackButton.native.js:
```diff
@@ -0,0 +1,59 @@
+     * @returns {ReactElement}
+     */
+    render() {
+        const cssPostfix = _.startCase(_.toLower(Platform.OS));
```
the OS is always given in lowercase

saghul commented on react/features/app-settings/components/BackButton.native.js:
```diff
@@ -0,0 +1,59 @@
+                onPress = { this.props.onPress }>
+                <Icon
+                    name = {
+                        Platform.OS === 'android'
```
make this a constant instead of computing it on every render

saghul commented on react/features/app-settings/components/FormRow.native.js:
```diff
@@ -39,7 +47,7 @@ type Props = {
  * on a form. The component should have exactly one child component.
  */
-class FormRow extends Component<Props> {
+class FormRow extends Component<Props, *> {
```
ditto

saghul commented on react/features/app-settings/components/styles.js:
```diff
@@ -4,32 +4,49 @@ import {
+    /**
+    * The back button style on Android platform.
+    */
+    backIconAndroid: {
```
I think this could be simplified using `Platform.select`: https://facebook.github.io/react-native/docs/platform-specific-code.html

saghul commented on react/features/app-settings/functions.native.js:
```diff
@@ -0,0 +1,53 @@
+
+    return (
+        Platform.OS === 'ios'
+        && ((SCREEN_HEIGHT === IPHONEX_HEIGHT && SCREEN_WIDTH === IPHONEX_WIDTH)
```
style: these need to be indented one extra level

saghul commented on react/features/app-settings/functions.native.js:
```diff
@@ -0,0 +1,53 @@
+* FIXME: Settings screen uses a custom implementation of safe views
+* to avoid empty areas on the screen when there should be a header rendered.
+* In long term we need to use a generic solution to handle "safe view
+* areas" for various devices. It can be React 0.50's {@code SafeAreaView} or
```
I didn't know it already existed in RN! why are we not using it then? wouldn't it be much simpler than what we are doing here now?

zbettenbuk replied:
I posted a note on Stride about how much I don't like how it renders the screen (too much empty space). Let's talk about it later, when we implement something more generic

saghul replied:
ok

saghul commented:
@zbettenbuk the top padding seems incorrect on non-X iPhones in landscape mode:

![img_0018](https://user-images.githubusercontent.com/317464/34767646-ef84b8d2-f5f8-11e7-9438-172141a94d82.PNG)

saghul commented on react/features/app-settings/components/AppSettings.native.js:
```diff
@@ -23,55 +28,76 @@ import { translate } from '../../base/i18n';
      */
     render() {
-        const { t } = this.props;
+        const { _profile, t } = this.props;
```
deconstructing _profile may also be worth it here

zbettenbuk replied:
I'm not sure it is. Reading the code, "profile.defaultURL" is something I understand immediately, while a variable simply named "defaultURL" may have multiple meaning.

saghul commented on react/features/app-settings/functions.native.js:
```diff
@@ -0,0 +1,73 @@
+*/
+export function isIPad() {
+    const {
+        height: SCREEN_HEIGHT,
```
why the capitalized values?

saghul commented on react/features/app-settings/functions.native.js:
```diff
@@ -0,0 +1,73 @@
+*
+* @returns {boolean}
+*/
+export function isIPad() {
```
this looks like something which belongs in base/react/platform

zbettenbuk replied:
I think the comment applies this too: these pieces of logic (actually maybe the entire file) needs to be refactored when we have a generic SafeArea solution, and since no other feature uses this method yet, I'd leave it here now. Also, I'd still prefer a hardware-info based solution replacing these.

saghul replied:
There are other reasons why we might want to test for a specific device. This is related to the platform itself, so it certainly belongs there. If we end up removing it great, if not, it will already be in the right place. Since this code is being added now, let's add it in the right place.

paweldomas replied:
I would also vote for moving outside. Nobody will know that this thing exists here and will duplicate it eventually.

saghul commented on react/features/app-settings/functions.native.js:
```diff
@@ -0,0 +1,73 @@
+*
+* @returns {boolean}
+*/
+function _isIPhoneX() {
```
ditto

saghul commented on react/features/app-settings/functions.native.js:
```diff
@@ -0,0 +1,73 @@
+
+    return Platform.OS === 'ios' && (
+        Math.max(SCREEN_HEIGHT, SCREEN_WIDTH)
+        / Math.min(SCREEN_HEIGHT, SCREEN_WIDTH)) < 1.6;
```
style: one more indent level

saghul commented on react/features/app-settings/components/BackButton.native.js:
```diff
@@ -0,0 +1,56 @@
+                onPress = { this.props.onPress }>
+                <Icon
+                    name = {
+                        isAndroid ? 'arrow_back' : 'navigate_before'
```
extract this computation to a constant at the top level, there is no reason to compute it every time

zbettenbuk commented on lang/main.json:
```diff
@@ -541,8 +541,8 @@
         "email": "Email",
+        "defaultURL": "Server URL",
         "header": "Settings",
-        "serverURL": "Server URL",
```
Can't really see, what's the problem with it?

saghul replied:
all language files would need to be updated. let's leave this change out

paweldomas replied:
If we change the key then translation will be broken, because those translation will be for the old key. But it depends if this's been translated already. Given how fresh the profileModal is, I think we can change it.

paweldomas replied:
@saghul currently it's only in main.json, but I don't know which one is better so I don't mind if it stays

saghul replied:
Yeah, see my other comment though.

paweldomas replied:
Yes, it's the server URL. "default URL" feels more like a whole thing including room name or whatever.

saghul commented on react/features/app-settings/components/BackButton.native.js:
```diff
@@ -0,0 +1,54 @@
+import { Icon } from '../../base/font-icons';
+import { Platform } from '../../base/react';
+
+const ICON = Platform.OS === 'android' ? 'arrow_back' : 'navigate_before';
```
please add a comment

saghul replied:
also a more descriptive name such as BACK_ICON_NAME sounds more appropriate IMHO

paweldomas commented on react/features/app-settings/components/AppSettings.native.js:
```diff
@@ -30,48 +45,58 @@ class AppSettings extends AbstractAppSettings {
+                        styles.headerContainer,
+                        this._getSafetyPadding()
+                    ] } >
+                    <BackButton
```
is the Android hardware back button handled by the profile settings screen ?

zbettenbuk replied:
Yes

saghul commented on react/features/app-settings/functions.native.js:
```diff
@@ -0,0 +1,73 @@
+*/
+function _isIPhoneX() {
+    const {
+        height: screenHeight,
```
I don't think the rename adds anything here, it's used right below and the context is clear enough.

zbettenbuk commented on react/features/base/react/functions.native.js:
```diff
@@ -0,0 +1,37 @@
+*
+* @returns {boolean}
+*/
+export function isIPad() {
```
Based on this page, yes: https://www.lifewire.com/list-of-ipad-models-and-generations-1994232

Btw there is a threshold, calculating that, as all iPad models (so far) has a ratio 1:1.33, so 1.6 seems to be a safe limit (with status bar and everything included).

saghul replied:
Ack

zbettenbuk commented:
@saghul thanks, fixed all, left comment on one.

About the section headers: yes, Lyubo suggested to have something, but the styling is mine, so that may need additional review from @yanas

saghul commented:
> About the section headers: yes, Lyubo suggested to have something, but the styling is mine, so that may need additional review from @yanas

No problem, we can improve that a bit later, I just wanted to know where it came from 👍

zbettenbuk commented:
@saghul I'm checking the tests, but weird, they are. The same test fails for Hristo's other PR.

saghul commented:
@damencho ping, can you help :-)

damencho commented:
Yep, just disabled the test and will look at it later, apparently, there is some problem between it and FF.
Also edited the template index file to have the correct link to chrome-2-chrome tests and to firefox-2-chrome tests, will take effect on new ran PR tests.

lyubomir commented on react/features/base/profile/middleware.js:
```diff
@@ -2,7 +2,10 @@
 import { PROFILE_UPDATED } from './actionTypes';
 import MiddlewareRegistry from '../redux/MiddlewareRegistry';
```
Please notice that we're importing from the redux feature in multiple lines!

````javascript
import MiddlewareRegistry from '../redux/MiddlewareRegistry';
import { toState } from '../redux';
````

That's why we group imports in the order third-party, other/outside feature, this/inside feature and separate the groups with empty lines. Let's please follow the coding style!

I'm taking care of this one.

lyubomir commented on react/features/base/profile/middleware.js:
```diff
@@ -31,13 +34,11 @@ MiddlewareRegistry.register(store => next => action => {
  */
 function _updateLocalParticipant(store) {
+    const localParticipant = getLocalParticipant(toState(store));
     const profile = getProfile(toState(store));
```
Please note that `toState(store)` is executed twice in the two lines above! That's very likely the result of copy-pasting that that's why copy-pasting is discouraged by our coding style.

I'm taking care of this one.

lyubomir commented on react/features/base/profile/middleware.js:
```diff
@@ -31,13 +34,11 @@ MiddlewareRegistry.register(store => next => action => {
-        name: profile.displayName
-    };
+    localParticipant.email = profile.email;
+    localParticipant.name = profile.displayName;
```
Ouch! Did we just modify the redux state here?

zbettenbuk replied:
Got the feedback to do it like this - assumed that I was asked to do this since this is only a copy of the store (or was it your point back then @saghul ? not entirely sure because we didn't discuss it). Originally the object was deconstructed and a new instance created (see above).

saghul replied:
Probably misread, my bad! Thanks for catching that @lyubomir !

lyubomir commented on react/features/base/react/functions.native.js:
```diff
@@ -0,0 +1,37 @@
+* @returns {boolean}
+*/
+export function isIPhoneX() {
+    const { height, width } = Dimensions.get('window');
```
I don't like `isIPad` and `isIPhoneX`. We're in freaking react-native which allows us to write native source code to query iOS what the device is and we're writing some obscure source code involving the app's window dimensions and aspect rations?! Why are we OK with it, @saghul?

zbettenbuk replied:
Ok, so two things to explain here:

- The point of using an implementation like this is to minimise the blank area around screens while properly handle the iPhoneX top notch and iPhone's status bar (that behaves differently on iPad). I could have used React's new SafeArea component, but with screens that have a header (or similar, native looking screens) it renders a bad looking empty area around, that this implementation avoids.
- In another discussion with Saul we talked about introducing a new native module that would do similar things by getting these info from the device but it turned out that it doesn't solve all our problems so we decided not to introduce a new native module if it only gives us half of the solution.

saghul replied:
We made an attempt to use SafeArea as is, but IMHO it needs to be a larger change which other components such as WelcomePage and Conference also use, so I conceded going this route since the broken code was already in.

lyubomir commented on react/features/app-settings/components/AbstractAppSettings.js:
```diff
@@ -10,6 +10,16 @@ import { getProfile, updateProfile } from '../../base/profile';
 
+    /**
+    * The current aspect ratio of the screen.
+    */
```
I'd like to please attract your attention to the formatting of the multiline comments. I presume it's an editor setting so could you please make sure you comply with our coding style?

We don't do:
````javascript
/**
*
*/
````

We do:
````javascript
/**
 *
 */
````
